### ISSUE TITLE
Spaces in file paths used by the resolver are being encoded

### ISSUE THREAD

**Issue by tillig** [tillig/vscode-spectral#6]
Based on something found in #5, after adding the file resolver to deal with file references, a new issue was discovered where paths that have spaces in them are being encoded to `%20`.

**Comment by anteprimorac** [comment]
@tillig it looks like prepending path with `file://` resolves the issue.

**Comment by tillig** [comment]
Is that something you do in your spec, or something I need to do in the linter setup?

**Comment by anteprimorac** [comment]
@tillig in the spec. I assume it's overriding detection of which resolver to use. Probably the main issue is that if you left out schema part from the URI, it thinks that it's HTTP URI.

**Comment by tillig** [comment]
Hmmm. And that doesn't happen with the Spectral CLI?

I'm finding the internals of Spectral to be less than obvious, but I am [following the same set of default resolvers](https://github.com/stoplightio/spectral/blob/71de553244331531a0d4df1e3452158fd02e9f7a/src/resolvers/http-and-file.ts#L8) that they use for the CLI.

Which is to say, [if you don't specify a custom resolver](https://github.com/stoplightio/spectral/blob/71de553244331531a0d4df1e3452158fd02e9f7a/src/cli/services/linter/linter.ts#L34) (which my extension doesn't currently support anyway), then it defaults to [a combined file and HTTP resolver](https://github.com/stoplightio/spectral/blob/71de553244331531a0d4df1e3452158fd02e9f7a/src/cli/services/linter/utils/getResolver.ts#L14) that looks [like this](https://github.com/stoplightio/spectral/blob/develop/src/resolvers/http-and-file.ts). That's [exactly what I'm doing, too](https://github.com/tillig/vscode-spectral/blob/7410b6af7c084cea09a8ec1c83ca42ccc6d1b8f2/server/resolver.ts#L9).

Let me see if I can repro it with the Spectral CLI. If it happens there, too, then maybe there's a bug in Spectral that needs to be reported.

**Comment by tillig** [comment]
Doesn't look like it's happening in the Spectral CLI so let me see what's going on. Maybe it's some setting I'm missing.

**Comment by tillig** [comment]
This is going to be a tougher one to solve. I'd definitely be open to some help.

There's a difference in how the VS Code mechanism is calling the linter and how the CLI executes the linter. With the CLI, you pass in the name of the file(s) to lint and the file reading/processing happens inside the Spectral process. With VS Code, we have the document in the editor already read in and use that text (which allows you to lint things you haven't saved yet, for example, which are just in-memory).

I'm struggling to see how this is precisely different, though the CLI linter [basically starts here](https://github.com/stoplightio/spectral/blob/71de553244331531a0d4df1e3452158fd02e9f7a/src/cli/services/linter/linter.ts#L67). I'll have to see if I can step through this and figure it out.

That said... this isn't my day job, to be honest, so I can't promise a super immediate fix. If someone can help me figure out what I'm doing wrong, that'd be sweet.

**Comment by anteprimorac** [comment]
@tillig I am happy to help. Will try to set up dev env and debug it today. I found that vscode follow the link feature is not working on file schema URIs.

**Comment by tillig** [comment]
So far I've traced it down to the [`ResolveRunner` class from `json-ref-resolver`](https://github.com/stoplightio/json-ref-resolver/blob/e2a5a52bdfc0c631d2dcdc3442c63e2775852842/src/runner.ts#L333) where it computes the absolute URI for a relative path by doing some string concatenation against the `baseUri` of the runner.

By the time it gets here, the `baseUri` has already been processed to remove the `file://` protocol, but it has not be URL-decoded. Somewhere along the way it needs to be decoded but I'm not sure where that would happen yet.

**Comment by anteprimorac** [comment]
@tillig thanks for sharing that. Sorry but I didn’t have time today to spin up the dev env and debug it, but I think tomorrow will have time to do it. I will let you know if I find something.

**Comment by tillig** [comment]
I think I found the bug.

When `ResolveRunner` gets the base URI it checks to see if it's a file URI, then if it is, [it runs through `toFSPath` on the value and then creates a URI out of it](https://github.com/stoplightio/json-ref-resolver/blob/e2a5a52bdfc0c631d2dcdc3442c63e2775852842/src/runner.ts#L62).

The way it figures out if it's a file URI is [if the scheme is `file://` _or_ if there's no scheme at all](https://github.com/stoplightio/json-ref-resolver/blob/e2a5a52bdfc0c631d2dcdc3442c63e2775852842/src/runner.ts#L536).

[`toFSPath` is just `normalize`](https://github.com/stoplightio/path/blob/208fd8cc0c9076ab886594650b7db2af4e27641b/src/toFSPath.ts#L1) from the Stoplight `path` library, which [does absolutely no correct URL decoding](https://github.com/stoplightio/path/blob/208fd8cc0c9076ab886594650b7db2af4e27641b/src/normalize.ts#L10) despite the incoming `file://` protocol item being a URL.

This seems like a bug in the way local file URLs are handled. If there's no scheme, it doesn't need to be URL-decoded; but if there is (and there _is in VS Code_) then the file URL needs to be decoded. Or, alternatively, if there's no protocol then the base URI needs to be updated with a scheme _and_ it needs to be URL encoded so all `file://` URLs are treated the same.

Without debugging into it, I'd put money down that the difference with the CLI is that the local `file://` URI isn't being encoded properly before getting passed on to the linter, so things "just work."

Let me see what I can do.

**Comment by tillig** [comment]
Yeah, doing some mangling to undo the `file://` URI before starting the linter seems to work. Let me make sure this works on Windows, too, and I'll get a fix out.

**Comment by tillig** [comment]
Looks like the `json-ref-resolver` library is deprecated and eventually Spectral will move to `json-schema-ref-parser` [which doesn't have this issue - it correctly URL decodes things](https://github.com/APIDevTools/json-schema-ref-parser/blob/4fa22bcea67f2039c7a88e57cae1c4cf089c224b/lib/util/url.js#L188).

**Comment by anteprimorac** [comment]
@tillig awesome work. Thanks for doing this.
